docs/source/triggers/index.rst

```diff
@@ -125,159 +111,316 @@ Like ``Scenario`` or ``Update``, you can use ``do`` notation and
+well as other debug information.
+
+Diversion: Updating ``Message``
+===============================
+
+Before we can make our Trigger more useful, we need to think a bit more about
+what it is supposed to do. For example, we don't want to respond to ``bob``'s
+own messages. We also do not want to send messages when we have not received
+any.
+
+In order to start with something reasonably simple, we're going to set the rule as
+
+  *if* the last message we can see was *not* sent by ``bob``, *then* we'll send
+  ``"Please, tell me more about that."`` to whoever sent the last message we
+  can see.
+
+This raises the question of how we can determine which message is the last one,
+given the current structure of a message. In order to solve that, we need to
+add a ``Time`` field to ``Message``, which can be done by editing the
+``Message`` template in ``daml/User.daml`` to look like:
+
+.. code-block:: daml
+
+    template Message with
+      sender: Party
+      receiver: Party
+      content: Text
+      receivedAt: Time
+    where
+      signatory sender, receiver
+
+This should result in Daml Studio reporting an error in the ``SendMessage``
+choice, as it now needs to set the ``receivedAt`` field. Here is the updated
+code for ``SendMessage``:
+
+.. code-block:: daml
+
+    -- New definition for SendMessage
+        nonconsuming choice SendMessage: ContractId Message with
+            sender: Party
+            content: Text
+          controller sender
+          do
+            assertMsg "Designated user must follow you back to send a message" (elem sender following)
+            now <- getTime
+            create Message with sender, receiver = username, content, receivedAt = now
+
+The ``getTime`` action (`doc </daml/stdlib/Prelude.html#function-da-internal-lf-gettime-99334>`__)
+returns the time at which the command was received by the sandbox. In more
+sensitive applications, this may not be sufficiently reliable, as transactions
+may be processed in parallel (so "received at" timestamp order may not match
+actual transaction order), and in distributed cases dishonest participants may
+fudge this value. It's good enough for this example, though.
+
+Now that we have a field to sort on, and thus a way to identify the *latest*
+message, we can turn our attention back to our trigger code.
+
+AutoReply
+=========
+
+Open up the trigger code again (``daml/ChatBot.daml``), and change it to:
+
+.. code-block:: daml
+
+    module ChatBot where
+
+    import qualified Daml.Trigger as T
+    import qualified User
+    import qualified DA.List.Total as List
+
+    autoReply : T.Trigger ()
+    autoReply = T.Trigger
+      { initialize = pure ()
+      , updateState = \_ -> pure ()
+      , rule = \p -> do
+          message_contracts <- T.query @User.Message
+          let messages = map snd message_contracts
+          debug $ "Messages so far: " <> show (length messages)
+          let lastMessage = List.last $ List.sortOn (.receivedAt) messages
```

Comment:
definitely not for this PR but we should have a `maximumOn`.

Reply:
Done in #10527; this has been rebased on now uses `maximumOn`.